doc/guides/bluetooth/bluetooth-arch.rst

```diff
@@ -0,0 +1,539 @@
+connection object through the return value of the
+:cpp:func:`bt_conn_create_le` API.
+
+Security
+========
+
+To acheive a secure relationship between two Bluetooth devices a process
+called pairing is used. This process can either be triggered implicitly
+through the security properties of GATT services, or explicitly using
+the :cpp:func:`bt_conn_security` API on a connection object.
+
+To acheive a higher security level, and protect against
+Man-In-The-Middle (MITM) attacks, it is recommended to use some
+out-of-band channel during the pairing. If the devices have a sufficient
+user interface this "channel" is the user itself. The capabilities of
+the device are registered using the :cpp:func:`bt_conn_auth_cb_register`
+API.  The :c:type:`bt_conn_auth_cb` struct that's passed to this API has
+a set of optional callbacks that can be used during the pairing - if the
+device lacks some feature the corresponding callback may be set to NULL.
+For example, if the device does not have an input method but does have a
+display, the ``passkey_entry`` and ``passkey_confirm`` callbacks would
+be set to NULL, but the ``passkey_display`` would be set to a callback
+capable of displaying a passkey to the user.
+
+Depending on the local and remote security requirements & capabilities,
+there are four possible security levels that can be reached:
+
+    :cpp:enumerator:`BT_SECURITY_LOW`
+        No encryption and no authentication.
+
+    :cpp:enumerator:`BT_SECURITY_MEDIUM`
+        Encryption but no authentication (no MITM protection).
+
+    :cpp:enumerator:`BT_SECURITY_HIGH`
+        Encryption and authentication using the legacy pairing method
+        from Bluetooth 4.0 and 4.1.
+
+    :cpp:enumerator:`BT_SECURITY_FIPS`
+        Encryption and authentication using the LE Secure Connections
+        feature available since Bluetooth 4.2.
+
+.. note::
+  Mesh has its own security solution through a process called
+  provisioning. It follows a similar procedure as pairing, but is done
+  using separate mesh-specific APIs.
+
+L2CAP
+=====
+
+L2CAP layer enables connection-oriented channels which can be enable with the
+configuration option: :option:`CONFIG_BT_L2CAP_DYNAMIC_CHANNEL`. This channels
+support segmentation and reassembly transparently, they also support credit
+based flow control making it suitable for data streams.
+
+Channels instaces are represented by the :cpp:class:`bt_l2cap_chan` struct which
+contains the callbacks in the :cpp:class:`bt_l2cap_chan_ops` struct to inform
+when the channel has been connected, disconnected or when the encryption has
+changed.
+In addition to that it also contains the ``recv`` callback which is called
+whenever an incoming data has been received. Data received this way can be
+marked as processed by returning 0 or using
+:cpp:func:`bt_l2cap_chan_recv_complete` API if processing is asynchronous.
+
+.. note::
+  The ``recv`` callback is called directly from RX Thread thus it is not
+  allowed to block.
```

Comment:
This isn't completely true. You can (and should) block for small periods of time if it simplifies the code. This is e.g. why it's possible to do `bt_hci_cmd_send_sync()` from within the RX thread.

Reply:
So I'd change this to something like:

"The recv callback is called from the RX thread, so long blocking operations should be avoided and all blocking operations which rely on the RX thread itself must not be done since it would result in a deadlock."

Reply:
I will update the text, but in case of bt_l2cap_chan.recv it is a callback on the application, well appart from fixed channels, so I guess it is useful to inform in what context they will get called.